### PR TITLE
Series of optimization for hashconsing

### PR DESCRIPTION
**Pull request by ppedrot**

### CODE REVIEW

**Comment by ppedrot** [comment]
Bench: https://github.com/coq/coq/pull/14804.

File: clib/hashcons.ml [hunk 1]
```diff
@@ -65,8 +65,11 @@ module Make (X : HashconsedType) : (S with type t = X.t and type u = X.u) =
       (tab, u)
 
     let hcons (tab, u) x =
-      let y = X.hashcons u x in
-      Htbl.repr (X.hash y) y tab
+      let h = X.hash x in
+      if Htbl.mem h x tab then x
+      else
+        let y = X.hashcons u x in
+        Htbl.repr h y tab
```
Comment (Janno) [inline clib/hashcons.ml:72]: Two questions:
1. We are performing a total of two lookups when Htbl.mem returns false. Could that be part of why this PR can be slower than master?
2. If Htbl.mem succeeds then you return `x` whereas the original code would have returned the representative of `x` in the hashmap. Wouldn't this lead to less sharing?

Comment (Janno) [reply clib/hashcons.ml:72]: Oh, I think my second question doesn't make sense. Your implementation of `mem` actually checks for physical equality so returning `x` is also what would have happened in the previous version of this code.

Comment (ppedrot) [reply clib/hashcons.ml:72]: To answer your first question, maybe but I am not sure. I've been conducing local experiments and I don't know what to think. I'll retry with a refined version of this PR and see what happens.

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬──────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]   │   mem faults    │
│                             │                         │                                       │                                       │                          │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD   PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│      coq-mathcomp-character │  145.27   148.55  -2.21 │   403321511364    412918562356  -2.32 │   560398370753    576819096012  -2.85 │  732088   709388    3.20 │   0    0    nan │
│                   coq-color │  475.52   485.82  -2.12 │  1328108439968   1355196260296  -2.00 │  1658830486366   1700912265364  -2.47 │ 1159888  1155700    0.36 │   0    0    nan │
│       coq-mathcomp-fingroup │   47.24    48.20  -1.99 │   131109640591    133309533886  -1.65 │   182513035470    186830469778  -2.31 │  480308   482640   -0.48 │   0    0    nan │
│       coq-mathcomp-solvable │  166.22   169.19  -1.76 │   462957208575    470719901898  -1.65 │   652703520493    665623824842  -1.94 │  693652   691652    0.29 │   0    0    nan │
│               coq-perennial │ 7188.24  7279.62  -1.26 │ 20016335157679  20274419847043  -1.27 │ 30676094352148  30914786496847  -0.77 │ 3010416  2979540    1.04 │   0    0    nan │
│                coq-coqprime │  329.53   333.67  -1.24 │   914372421354    926732655061  -1.33 │  1597705362548   1595120252097   0.16 │  926044   913984    1.32 │   0    0    nan │
│        coq-mathcomp-algebra │  137.89   139.56  -1.20 │   383091008927    387766392136  -1.21 │   497006547044    502561170498  -1.11 │  555188   557120   -0.35 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1032.26  1043.71  -1.10 │  2875422337621   2906213580571  -1.06 │  4574526747982   4597638319303  -0.50 │  907464   913352   -0.64 │   0    0    nan │
│                   coq-verdi │   98.13    99.16  -1.04 │   270100374801    271682192036  -0.58 │   379815787709    379597787997   0.06 │  542948   545228   -0.42 │   0    0    nan │
│  coq-performance-tests-lite │ 1390.64  1404.28  -0.97 │  3878450300446   3921933778265  -1.11 │  6325087408949   6331645894355  -0.10 │ 2085028  2085096   -0.00 │   0    0    nan │
│                  coq-stdlib │  411.40   415.39  -0.96 │  1149770735739   1161877843310  -1.04 │  1465856665818   1470063438395  -0.29 │  598120   597556    0.09 │   0    0    nan │
│                  coq-geocoq │ 1316.94  1328.77  -0.89 │  3674076185330   3702420997646  -0.77 │  5427983797518   5422360037386   0.10 │ 1023444   980248    4.41 │   0    0    nan │
│                coq-bedrock2 │  301.37   303.99  -0.86 │   836113042498    843273171969  -0.85 │  1483154938542   1492860236890  -0.65 │  799992   800140   -0.02 │   0    0    nan │
│              coq-coquelicot │   68.57    69.16  -0.85 │   187832905174    188920211177  -0.58 │   241145595968    241330967446  -0.08 │  557820   558844   -0.18 │   0    0    nan │
│                    coq-corn │ 1455.76  1467.80  -0.82 │  4052059790667   4086787874304  -0.85 │  5853514537778   5874479890188  -0.36 │  801140   823024   -2.66 │   0    0    nan │
│                   coq-flocq │  146.87   148.04  -0.79 │   407270532580    410479049540  -0.78 │   506330167550    506370115512  -0.01 │  810276   810500   -0.03 │   0    0    nan │
│                 coq-bignums │   55.67    56.10  -0.77 │   153855113224    154917615650  -0.69 │   207314140321    207208496775   0.05 │  463268   463412   -0.03 │   0    0    nan │
│              coq-verdi-raft │ 1094.18  1102.58  -0.76 │  3047946416195   3070280974784  -0.73 │  4319951578256   4308082866728   0.28 │  915976   916556   -0.06 │   0    0    nan │
│             coq-lambda-rust │ 1263.81  1272.21  -0.66 │  3517145479082   3540337630593  -0.66 │  5086435745013   5084878077353   0.03 │ 1111176  1108052    0.28 │   0    0    nan │
│      coq-mathcomp-ssreflect │   53.45    53.77  -0.60 │   145855939409    147679431311  -1.23 │   189295859749    191577063601  -1.19 │  527204   530184   -0.56 │   0    0    nan │
│       coq-engine-bench-lite │  290.08   291.33  -0.43 │   816932821296    820969092870  -0.49 │  1498502524656   1496055520017   0.16 │ 1293852  1411696   -8.35 │   0    0    nan │
│             coq-fiat-crypto │ 4839.42  4858.69  -0.40 │ 13441081337113  13497139642170  -0.42 │ 23729616068257  23774783502778  -0.19 │ 2219008  2228908   -0.44 │   0    0    nan │
│               coq-fourcolor │ 2538.14  2543.98  -0.23 │  7075342667778   7092422767576  -0.24 │ 12241510160499  12238537940404   0.02 │  797416   732696    8.83 │   0    0    nan │
│            coq-fiat-parsers │  583.08   583.31  -0.04 │  1628589874258   1632623462418  -0.25 │  2493299677006   2492584802453   0.03 │ 3485536  3485560   -0.00 │   0    0    nan │
│          coq-mathcomp-field │  218.29   218.34  -0.02 │   607230897078    607517331174  -0.05 │   923414322417    918946482369   0.49 │  660800   658968    0.28 │   0    0    nan │
│                coq-rewriter │  685.88   685.66   0.03 │  1902435599640   1903549864236  -0.06 │  2933442276814   2931375137675   0.07 │ 1073932  1073900    0.00 │   0    0    nan │
│            coq-math-classes │  172.67   172.54   0.08 │   480340118969    482795410254  -0.51 │   641931106698    641207597230   0.11 │  482144   482000    0.03 │   0    0    nan │
│                    coq-hott │  298.84   298.11   0.24 │   822678836223    820959213809   0.21 │  1289566500433   1269308527376   1.60 │  570992   570916    0.01 │   0    0    nan │
│               coq-fiat-core │  108.58   108.20   0.35 │   306640549982    306447301184   0.06 │   416420167748    413086236611   0.81 │  480556   479944    0.13 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │ 1638.63  1631.05   0.46 │  4558119686094   4539255568503   0.42 │  7166624168831   7149812517820   0.24 │ 1073756  1073764   -0.00 │   0    0    nan │
│                    coq-core │  108.47   107.47   0.93 │   298867682654    298243085862   0.21 │   462708742274    462481923610   0.05 │  299692   299716   -0.01 │   0    0    nan │
│                 coq-unimath │ 4688.95  4645.35   0.94 │ 13051871767728  12925585683881   0.98 │ 25284343896142  24882183638060   1.62 │ 3267616  3712240  -11.98 │   0    0    nan │
│                 coq-coqutil │   60.09    59.35   1.25 │   166274661631    165826619159   0.27 │   219992551377    218566322676   0.65 │  601760   600000    0.29 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴──────────────────────────┴─────────────────┘
```
Maybe worth to be turned into an actual PR?

**Comment by JasonGross** [comment]
I'm curious about the distribution amongst files in coq-performance-tests-lite

**Comment by ppedrot** [comment]
I believe that only the two last commits really matter, I'll close this PR and reopen a fresh one with only those.

**Pull request closed without merging**